### PR TITLE
Embed Presto version in stacktraces

### PR DESCRIPTION
**Pull request by findepi**
Sometimes, users will just report a stacktrace, without telling which
Presto version it is for. Embed version information in exceptions
produced by Presto, so that the version is immediately know.

Example output
```
$ ./presto-product-tests/conf/docker/singlenode/compose.sh run application-runner /docker/volumes/conf/docker/files/presto-cli.sh --debug --catalog hive --schema default --user hive
+ java -jar /docker/volumes/presto-cli/presto-cli-executable.jar --server presto-master:8080 --debug --catalog hive --schema default --user hive
presto:default> select * from x;
Query 20181012_121120_00002_k4ata failed: line 1:15: Table hive.default.x does not exist
com.facebook.presto.sql.analyzer.SemanticException: line 1:15: Table hive.default.x does not exist
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitTable(StatementAnalyzer.java:848)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitTable(StatementAnalyzer.java:258)
	at com.facebook.presto.sql.tree.Table.accept(Table.java:53)
	at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:270)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.analyzeFrom(StatementAnalyzer.java:1772)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:954)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuerySpecification(StatementAnalyzer.java:258)
	at com.facebook.presto.sql.tree.QuerySpecification.accept(QuerySpecification.java:127)
	at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:270)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:280)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:676)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.visitQuery(StatementAnalyzer.java:258)
	at com.facebook.presto.sql.tree.Query.accept(Query.java:94)
	at com.facebook.presto.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer$Visitor.process(StatementAnalyzer.java:270)
	at com.facebook.presto.sql.analyzer.StatementAnalyzer.analyze(StatementAnalyzer.java:244)
	at com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:72)
	at com.facebook.presto.sql.analyzer.Analyzer.analyze(Analyzer.java:64)
	at com.facebook.presto.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:184)
	at com.facebook.presto.execution.SqlQueryExecution$SqlQueryExecutionFactory.createQueryExecution(SqlQueryExecution.java:721)
	at com.facebook.presto.execution.SqlQueryManager.createQueryInternal(SqlQueryManager.java:453)
	at com.facebook.presto.execution.SqlQueryManager.lambda$createQuery$3(SqlQueryManager.java:386)
	at com.facebook.presto.$gen.Presto_0_212_94_g1fabb82____20181012_121051_1.run(Unknown Source)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/version/EmbedVersion.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    private MethodHandle runnableConstructor;
+
+    @Inject
+    public EmbedVersion(ServerConfig serverConfig)
+    {
+        this.builtInVersion = new ServerConfig().getPrestoVersion();
+        this.configuredVersion = serverConfig.getPrestoVersion();
+    }
+
+    @PostConstruct
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/version/EmbedVersion.java]: why not in constructor?

File: presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java [hunk 2]
```diff
@@ -48,6 +48,7 @@
 import com.facebook.presto.sql.tree.Statement;
 import com.facebook.presto.transaction.TransactionManager;
 import com.facebook.presto.util.StatementUtils;
+import com.facebook.presto.version.EmbedVersion;
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: Could you attach an example stacktrace?

Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java]: Sure. I pasted it in https://github.com/prestodb/presto/pull/11700#issue-222447112

File: presto-main/src/main/java/com/facebook/presto/version/EmbedVersion.java (hunk 1 above)
Comment (findepi) [reply presto-main/src/main/java/com/facebook/presto/version/EmbedVersion.java]: I think this is better place guice-wise. i don't have strong opinion though.

**Comment by findepi** [comment]
> We could also embed version into the main thread (the one that loads configuration) and execution threads.

Yes, i am sure one place is not enough. I will try to find other appropriate places, but if you have suggestions, that would be helpful.

**Comment by kokosing** [comment]
> Yes, i am sure one place is not enough. I will try to find other appropriate places, but if you have suggestions, that would be helpful.

We can always do it as we go.

**Comment by wenleix** [comment]
This doesn't help with stacktrace that doesn't include bytecode generated class though.

Does it make sense to add version to QueryID? @martint , @dain , @electrum , @raghavsethi

**Comment by findepi** [comment]
> This doesn't help with stacktrace that doesn't include bytecode generated class though.

Obviously. I need to identify all relevant "execution roots". I have 2 so far.

> Does it make sense to add version to QueryID?

@wenleix i thought about this as well. We can add there as well, if you don't object query id to be a little longer.

**Comment by findepi** [comment]
@sopel39 @kokosing let's assume i'm done.

**Comment by nezihyigitbasi** [comment]
Changing the format of the query id to include the version may break tools depending on that.

**Comment by dain** [comment]
@wenleix The query was designed to be a short, consistent, format containing a lot of semantic meaning.  On the other hand, the version can be an arbitrarily long string containing virtually any text, so I think it would be more difficult to use the queryId if we added the version.

**Comment by dain** [comment]
As future work, I'd like to have a debugging tool that people can use to capture cluster/query information to speedup support.  Of course, the information would need to be carefully chosen to not include sensitive details.

**Pull request merged**